hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/HoodieSparkMergeHandleV2.java

```diff
@@ -0,0 +1,413 @@
+          = (HoodieFileGroupReaderIterator<InternalRow>) fileGroupReader.getClosableIterator();
+      StructType sparkSchema = AvroConversionUtils.convertAvroSchemaToStructType(readerSchema);
+      while (recordIterator.hasNext()) {
+        InternalRow row = recordIterator.next();
+        HoodieKey recordKey = new HoodieKey(
+            row.getString(HoodieRecord.COMMIT_SEQNO_METADATA_FIELD_ORD),
+            row.getString(HoodieRecord.PARTITION_PATH_META_FIELD_ORD));
+        HoodieSparkRecord record = new HoodieSparkRecord(recordKey, row, sparkSchema, false);
+        Option recordMetadata = record.getMetadata();
+        if (!partitionPath.equals(record.getPartitionPath())) {
+          HoodieUpsertException failureEx = new HoodieUpsertException("mismatched partition path, record partition: "
+              + record.getPartitionPath() + " but trying to insert into partition: " + partitionPath);
+          writeStatus.markFailure(record, failureEx, recordMetadata);
+          continue;
+        }
+        try {
+          writeToFile(recordKey, record, readerSchema, config.getPayloadConfig().getProps(), preserveMetadata);
+          writeStatus.markSuccess(record, recordMetadata);
+        } catch (Exception e) {
+          LOG.error("Error writing record  " + record, e);
+          writeStatus.markFailure(record, e, recordMetadata);
+        }
+
+      }
+      // The stats of inserts, updates, and deletes are updated once at the end
+      HoodieFileGroupReaderStats stats = fileGroupReader.getStats();
+      this.insertRecordsWritten = stats.getNumInserts();
+      this.updatedRecordsWritten = stats.getNumUpdates();
+      this.recordsDeleted = stats.getNumDeletes();
+      this.recordsWritten = stats.getNumInserts() + stats.getNumUpdates();
+    } catch (IOException e) {
+      throw new HoodieUpsertException("Failed to compact file slice: " + fileSlice, e);
+    }
+  }
+
+  protected void writeToFile(HoodieKey key, HoodieSparkRecord record, Schema schema, Properties prop, boolean shouldPreserveRecordMetadata)
+      throws IOException {
+    // NOTE: `FILENAME_METADATA_FIELD` has to be rewritten to correctly point to the
+    //       file holding this record even in cases when overall metadata is preserved
+    MetadataValues metadataValues = new MetadataValues().setFileName(newFilePath.getName());
+    HoodieRecord populatedRecord = record.prependMetaFields(schema, writeSchemaWithMetaFields, metadataValues, prop);
+
+    if (shouldPreserveRecordMetadata) {
+      fileWriter.write(key.getRecordKey(), populatedRecord, writeSchemaWithMetaFields);
+    } else {
+      fileWriter.writeWithMetadata(key, populatedRecord, writeSchemaWithMetaFields);
+    }
+  }
+
+  @Override
+  public List<WriteStatus> close() {
+    try {
+      if (isClosed()) {
+        // Handle has already been closed
+        return Collections.emptyList();
+      }
+
+      markClosed();
+
+      writtenRecordKeys = null;
```

Comment:
This is useless for fg reader, it's used for collect left over inserts BTW for the old merge handle.

Reply:
I now extends the class from `HoodieMergeHandle` so all of these logic is removed or reused.